### ISSUE TITLE
Lazy loading in backend environment? (Deno)

### ISSUE THREAD

**Issue by aLemonFox** [usnistgov/h5wasm#69]
I have a backend server where I want to lazy load a dataset to limit my bandwidth usage. Based on #4 I tried to implement it like so:
```ts
try {
  const Modules = await h5wasm.ready;
  const { FS } = Modules;

  FS.createLazyFile("/", "temp.h5", signedUrl, true, false);
  const file = new h5wasm.File("current.h5");

  console.log(file);
} catch (err) {
  console.error(err);
}
```

But this doesn't seem to work. I think because my environment is not setup properly as it is not web-based. Is there anything else needed to configure lazy url based access in Deno (and Node)?

**Comment by bmaranville** [comment]
I'm not completely sure - the createLazyFile function seems to have been written specifically for the browser context, as it uses 
`new XMLHttpRequest...` in the code (see https://github.com/emscripten-core/emscripten/blob/53f661cb11ba849403c060b97208f88775484d98/src/library_fs.js#L1678C1-L1678C42)

You might be able to use this shim library to get it to work: https://www.npmjs.com/package/xmlhttprequest

**Comment by aLemonFox** [comment]
That could work. Do you have an idea on how to patch it as I don't think the FS comes bundled with this lib right?

**Comment by bmaranville** [comment]
deno apparently has a solution for xhr, and I was able to get past your first error:
```node
> let xhr = await import("https://deno.land/x/xhr@0.3.1/mod.ts")
undefined
> try {
  const Modules = await h5wasm.ready;
  const { FS } = Modules;

  FS.createLazyFile("/", "current.h5", signedUrl, true, false);
  const file = new h5wasm.File("current.h5");

  console.log(file);
} catch (err) {
  console.error(err);
}
Cannot do synchronous binary XHRs outside webworkers in modern browsers. Use --embed-file or --preload-file in emcc
undefined
```

More investigation required... I don't know the current status of web workers in Deno.

**Comment by bmaranville** [comment]
It looks like Deno doesn't support synchronous fetch/xhr even in a web worker.  I think the sync flag is required for the createLazyFile implementation in Emscripten, and it's the test for that flag that is failing and throwing the current error.  I don't see any way to do a synchronous fetch in Deno, and I don't see any way to do an async file read in HDF5 (without writing a new Virtual File Driver), so I'm not sure if there's an easy path forward.  If you find something, please let me know!

**Comment by aLemonFox** [comment]
I havn't found a way to make it work using your lib, but made a workaround using `gdal-async`. It has support for writing COG geotiff files (https://www.cogeo.org/) which also allow for on demand loading of the dataset.

Then using [geoblaze](https://github.com/GeoTIFF/geoblaze) and [georaster](https://github.com/GeoTIFF/georaster) I am able to slice the needed values of my dataset on demand.

**Comment by aLemonFox** [comment]
Well, there seems to be an interesting difference the way dependencies are handled in Deno. For example importing `geoblaze` using `npm:geoblaze` vs `https://esm.sh/geoblaze` result in different outcomes for worker based requests. Where the `esm` version does not work, the `npm:` import works fine.

**Comment by bmaranville** [comment]
Thanks for the tip... I'll check out JSR.

**Comment by bmaranville** [comment]
Is it possible the `npm:` import is using a different implementation of a web worker, instead of the one distributed with Deno?

**Comment by aLemonFox** [comment]
Yeah that seems like it, but I can't figure out how to change it. I've worked on another solution using some nodejs serverless functions to handle this as a service.
